src/librustc_mir/transform/simplify_cfg.rs

```diff
@@ -125,8 +150,7 @@ impl MirPass for SimplifyCfg {
         let mut changed = true;
         while changed {
             changed = self.simplify_branches(mir);
-            changed |= self.remove_goto_chains(mir);
-            self.remove_dead_blocks(mir);
+            changed |= self.merge_consecutive_blocks(mir);
```

Comment:
I think that doing `merge_consecutive_blocks` without a `remove_dead_blocks` pass beforehand is missing out on some optimizations. For example:

```
Entry -> B <- A
```

Since `A` is dead, `B` only has one "real" predecessor, and so can be merged into `Entry`. However, since dead blocks aren't removed until the afterwards, `merge_consecutive_blocks` sees `B` as having 2 predecessors and doesn't merge it into `Entry`.